### ISSUE TITLE
--merge-subtitle guesses language with a webservice

### ISSUE THREAD

**Issue by olof** [spaam/svtplay-dl#1262]
If I understand correctly, @iwconfig wrote the --merge-subtitles feature, and it is implemented as web service hosted on heroku. It does so by taking random samples from the subtitle and posts it to the web service that then returns some guess as to what language it is.

What is the reason for implementing it like this, and not integrating it in svtplay-dl itself? And is there no metadata we can use rather than guessing the language?

**Comment by iwconfig** [comment]
It was mostly a quick and easy solution to a problem I had. As far as I know, SVT Play does not provide metadata for this. It is mostly in Swedish anyway but lets not discriminate our fellow immigrants. UR Play does provide though (IIRC), while others might not. That's the main reason. You would have to implement it for each service which is a hassle and if one of them does not provide the metadata, the app conformity suffers.

The web service is solely my own, it's on my personal Heroku account. I was quite novice back then so I imagined it would be a hassle to include yet another two modules, as I was afraid my PR would not be as appreciated (a big deal for me personally, at the time :). Also, I just thought it would be fun to experiment with an API. Its not really a big deal to integrate it in svtplay-dl itself, partly because I know it is used quite often because I use the stathat API to count each call (IP's are **not** recorded ofc, it's only a visit counter).

It is quite fast when the web dyno is awake, and IMO not painfully slow when it is asleep either. Also the free dyno hours tend to last through out the month, only to run out during christmas/new year and the like but not always. Is any of this a concern? If so I can take a look at integrating it instead when I have time available.

One way could be to provide both: use metadata and if it's not available then use language detector either as a web service or directly implemented. This would be handy because I have this general fear that some metadata are unreliable. But then if a language detector is already available, even the reliable metadata is more or less redundant.

**Comment by olof** [comment]
Hello, thanks a lot for explaining the background! In my mind, it would be better for svtplay-dl to do its thing in isolation (well, we still rely on the service we target being available of course!). I don't think this is very important as such, especially since it is an opt-in code path, but it was not clear to me that a flag --merge-subtitle would result in network traffic and having it not do so would be better from a user expectation point of view.

Regarding the metadata: I tried to look at svtplay, and it seems you are mostly right in that they provide no/limited metadata about the subtitle. But I did notice that in the web player, they still list it as "Swedish [CC]". Digging a bit further (literally digging through megabytes of minified javascript <3), they do have some mapping function that maps iso6632-2 codes to "nice names", which a very limited set of languages available (`getReadableTextTrackName` in `client.bundle.xxxx.js`). That, I think, means that some where, the iso6632-2 code is available to the client, but it was not passed with the other video metadata it seems... Or it is hardcoded somewhere else. Ok, that made me tired, so I'll do something else now! :)

So, to sum it up: It should be possible to extract metadata, but as you point, not at all straight forward. Perhaps keep this issue open as a place to discuss such a change. Maybe you'll do it when you find time/motivation, maybe I do it, maybe nobody! Or somebody else, who knows. Open source ♡. (And to be honest, the onus is probably on me to change something that is working as is. Again, thanks for the background!)

**Comment by SYSophie** [comment]
> So, to sum it up: It should be possible to extract metadata, but as you point, not at all straight forward. Perhaps keep this issue open as a place to discuss such a change. Maybe you'll do it when you find time/motivation, maybe I do it, maybe nobody! Or somebody else, who knows. Open source ♡. (And to be honest, the onus is probably on me to change something that is working as is. Again, thanks for the background!)

Metadata for subtitles etc is presented in HLS-manifest

`#EXT-X-MEDIA:TYPE=SUBTITLES,URI="text/text-0.m3u8",GROUP-ID="text",LANGUAGE="sv",NAME="Svenska",DEFAULT=NO,AUTOSELECT=YES,CHARACTERISTICS="public.accessibility.transcribes-spoken-dialog,public.accessibility.describes-music-and-sound"`

**Comment by olof** [comment]
Ah, that makes sense! Thanks @SYSophie

**Comment by StaffanBetner** [comment]
One alternative is to use CLD2 https://github.com/CLD2Owners/cld2

**Comment by iwconfig** [comment]
@SYSophie I now remember that I, at the time, was looking into using that data from both DASH and HLS manifests, but IIRC I came to the conclusion they were not a reliable source. Svtplay aside, especially considering other sites might not provide it accurately, I thought using a translator was the better alternative. **It's not to say we couldn't utilize both, tough.**

I roughly looked at this today comparing svtplay and urplay.

- Svtplay really only has one subtitle (Swedish). I haven't found any videos with more than one subtitle yet.
(IIRC "Nyheter på lätt svenska" did some years ago, but not anymore).

- URPlay **does not** provide language metadata in their `playlist.m3u8` file.

At the moment, it seems there isn't a canonical way to solve this other than using a translator.

### If you look into this as well, maybe other services, please share your thoughts and findings
